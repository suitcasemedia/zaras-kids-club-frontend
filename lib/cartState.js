import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;
function CartStateProvider({ children }) {
  // this is its own custom provider

  const [cartOpen, setCartOpen] = useState(false);
  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// make a custom hook for acessing the cart local state

function useCart() {
  const all = useContext(LocalStateContext);
  return all;
}
export { CartStateProvider, useCart };
