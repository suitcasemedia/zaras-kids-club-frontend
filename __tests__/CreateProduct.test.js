import { screen, render, waitFor } from '@testing-library/react';
import Router from 'next/router';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import wait from 'waait';
import CreateProduct, {
  CREATE_PRODUCT_MUTATION,
} from '../components/CreateProduct';
import { fakeItem } from '../lib/testUtils';

const item = fakeItem();

jest.mock('next/router', () => ({
  push: jest.fn(),
}));

describe('<CreateProduct/>', () => {
  it('renders and matches snapshot', async () => {
    const { container } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('handles state updating', async () => {
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    // simulate someone filling out the form
    await userEvent.type(screen.getByPlaceholderText(/name/i), item.name);
    await userEvent.type(
      screen.getByPlaceholderText(/price/i),
      item.price.toString()
    );
    await userEvent.type(
      screen.getByPlaceholderText(/description/i),
      item.description
    );
    expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
  });

  it('creates an item when the form is submitted', async () => {
    const mocks = [
      {
        request: {
          query: CREATE_PRODUCT_MUTATION,
          variables: {
            name: item.name,
            description: item.description,
            image: '',
            price: item.price,
          },
        },
        result: {
          data: {
            createProduct: {
              ...fakeItem(),
              id: 'abc123',
              __typename: 'Product',
            },
          },
        },
      },
    ];

    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <CreateProduct />
      </MockedProvider>
    );
    // simulate someone filling out the form
    await userEvent.type(screen.getByPlaceholderText(/name/i), item.name);
    await userEvent.type(
      screen.getByPlaceholderText(/price/i),
      item.price.toString()
    );
    await userEvent.type(
      screen.getByPlaceholderText(/description/i),
      item.description
    );
    // mock the router
    Router.push = jest.fn();
    // submit the form
    await userEvent.click(screen.getByTestId('submitButton'));
    // wait for the form to submit
    await waitFor(() => wait(500));
    expect(Router.push).toHaveBeenCalled();
    expect(Router.push).toHaveBeenCalledWith({
      pathname: '/product/abc123',
    });
  });
});
