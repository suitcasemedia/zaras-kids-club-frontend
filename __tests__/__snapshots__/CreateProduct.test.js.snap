// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct/> renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 lnoXHv"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="name"
      >
        Image
        <input
          id="image"
          name="image"
          required=""
          type="file"
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="Price"
          type="number"
          value="0"
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="Description"
          type="text"
        />
      </label>
      <button
        data-testid="submitButton"
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;
